anchor/programs/sponsor-relayer/src/instructions/relay.rs

```diff
@@ -0,0 +1,98 @@
+    pub const COMPUTE_PRICE_DISCRIMINATOR: u8 = 3;
+    declare_id!("ComputeBudget111111111111111111111111111111");
+}
+
+#[derive(Accounts)]
+pub struct Relay<'info> {
+    #[account(mut)]
+    pub relayer_wallet: Signer<'info>,
+    #[account(has_one = relayer_wallet, has_one = sponsorship)]
+    pub relayer: Account<'info, Relayer>,
+    #[account(mut)]
+    pub sponsorship: Account<'info, Sponsorship>,
+    /// CHECK: instructions sysvar
+    #[account(address = tx_instructions::ID)]
+    pub instructions: UncheckedAccount<'info>,
+}
+
+pub fn relay(ctx: Context<Relay>) -> Result<()> {
+    let mut index = 0;
+    let mut signers = vec![];
+    let mut compute_limit = 0;
+    let mut compute_price = 0;
+    let mut init_accounts = vec![];
+
+    while let Ok(instr) = tx_instructions::load_instruction_at_checked(index, &ctx.accounts.instructions) {
+
+        instr.accounts.iter().for_each(|account| if account.is_signer {
+            if !signers.contains(&account.pubkey) {
+                signers.push(account.pubkey);
+            }
+        });
+
+        if instr.program_id == compute_budget::id() {
+            if instr.data[0] == compute_budget::COMPUTE_LIMIT_DISCRIMINATOR {
+                compute_limit = u64::from_le_bytes(instr.data[1..].try_into().unwrap());
+            } else if instr.data[0] == compute_budget::COMPUTE_PRICE_DISCRIMINATOR {
+                compute_price = u64::from_le_bytes(instr.data[1..].try_into().unwrap());
+            }
+        } else if instr.program_id == crate::id() {
+            //curr prog
+        } else if instr.program_id == anchor_spl::token::ID || instr.program_id == anchor_spl::token_2022::ID {
+            //spl token
+            // check if it's a transfer
+            if instr.data[0] == TRANSFER_CHECKED_DISCRIMINATOR {
+                if let Some(mint) = ctx.accounts.sponsorship.mint {
+                    require!(instr.accounts[1].pubkey == mint, SponsorRelayerError::InvalidMint);
+                }
+            } else if instr.data[0] == INITIALIZE_ACCOUNT_2_DISCRIMINATOR || instr.data[0] == INITIALIZE_ACCOUNT_3_DISCRIMINATOR {
+                init_accounts.push(instr.accounts[0].pubkey);
+            } else if instr.data[0] == SET_AUTHORITY_DISCRIMINATOR && instr.data[1] == CLOSE_AUTHORITY_TYPE {
+                require!(instr.data[3..] == ctx.accounts.relayer_wallet.key().to_bytes(), SponsorRelayerError::InvalidCloseAuthority);
+                if let Some(idx) = init_accounts.iter().position(|&x| x == instr.accounts[0].pubkey) {
+                    init_accounts.remove(idx);
+                } else {
+                    return Err(SponsorRelayerError::CloseAuthorityNotSet.into());
+                }
+            }
+        } else {
+            return Err(SponsorRelayerError::InvalidProgram.into());
+        }
+        
+        index += 1;
+    }
+
+    if init_accounts.len() > 0 {
+        return Err(SponsorRelayerError::CloseAuthorityNotSet.into());
```

Comment:
Is this the intended error?
I need a guide on all the checking above.

Reply:
yes, in this scenario there is a create token account without the respective close authority being set. I can go over this tomorrow on our sync